### PR TITLE
[Enhancement] support incremental scan ranges  deployment at BE side

### LINKED ISSUE #50196: To support incremental scan ranges deployment.
## Enhancement

In practice we have found that it can take a lot of time on the FE to get all the files that need to be scanned, and then send them down to the BE for execution. This is fine for native table, but not ideal for hive/iceberg/hudi/deltalake.

Ideally, the FE should start executing on the BE after scanning some of the files, while the FE continues to scan the rest of the files. This way, the whole execution process can be executed in parallel on both the FE and the BE, thus shortening the execution time.

Arch diagram:

image

two new session variables:
- enable_connector_incremental_scan_ranges=true (if to enable incremental scan ranges deployment)
- connector_incremental_scan_ranges_size=50  (if enabled, how many scan ranges delivered each round)

To achieve this goal, the following things need to be done. 
- [x] the FE should support incremental acquisition of scanned files.
  - https://github.com/StarRocks/starrocks/pull/49230
  - https://github.com/StarRocks/starrocks/pull/49352
  - https://github.com/StarRocks/starrocks/pull/49451
- [x] incremental deployment of files to the BE should be supported on the FE. 
  - https://github.com/StarRocks/starrocks/pull/50189
- [x] the BE should support incremental execution of files.
  - https://github.com/StarRocks/starrocks/pull/50254

### PR DESCRIPTION
**Pull request by dirtysalt**
## What I'm doing:

Compared to the FE modification, the BE modification is simpler, after receiving the request for incremental scan ranges
1. find the fragment context according to query_id + fragment_instance_id
2. find the corresponding morsel queue factory according to scan_node_id
3. morsel queue factory adds the incremental scan ranges to morsel queue

Fixes #50196 

### sf100 

default with file cache

|           |             |          |          |           |           |
| --------- | ----------- | -------- | -------- | --------- | --------- |
| **Query** | **Default** | **X=10** | **X=50** | **X=100** | **X=500** |
| **Q01**   | 3407        | 4464     | 3216     | 3274      | 3241      |
| **Q02**   | 354         | 362      | 330      | 339       | 340       |
| **Q03**   | 2126        | 2362     | 2031     | 2096      | 2048      |
| **Q04**   | 1322        | 1381     | 1051     | 1088      | 1071      |
| **Q05**   | 2707        | 2710     | 2538     | 2650      | 2596      |
| **Q06**   | 960         | 1906     | 1146     | 858       | 1133      |
| **Q07**   | 2495        | 2424     | 2333     | 2445      | 2440      |
| **Q08**   | 1562        | 1720     | 1572     | 1637      | 1597      |
| **Q09**   | 3965        | 4024     | 3605     | 3770      | 3624      |
| **Q10**   | 2420        | 2485     | 2567     | 2481      | 2423      |
| **Q11**   | 302         | 296      | 336      | 310       | 299       |
| **Q12**   | 1156        | 1307     | 1039     | 1000      | 1001      |
| **Q13**   | 2310        | 2463     | 2291     | 2301      | 2445      |
| **Q14**   | 947         | 965      | 911      | 967       | 961       |
| **Q15**   | 946         | 813      | 1157     | 924       | 1099      |
| **Q16**   | 1182        | 925      | 897      | 891       | 895       |
| **Q17**   | 879         | 1291     | 851      | 990       | 870       |
| **Q18**   | 7486        | 9813     | 7449     | 7663      | 7770      |
| **Q19**   | 1107        | 1255     | 1022     | 1045      | 1298      |
| **Q20**   | 1147        | 990      | 1292     | 1252      | 1120      |
| **Q21**   | 3369        | 3763     | 3455     | 3441      | 3518      |
| **Q22**   | 435         | 476      | 459      | 446       | 459       |
| **SUM**   | 42584       | 48195    | 41548    | 41868     | 42248     |

no file cache

|   |   |   |   |   |   |
|---|---|---|---|---|---|
|**Query**|**Default**|**X=10**|**X=50**|**X=100**|**X=500**|
|**Q01**|3167|4149|3209|3222|3257|
|**Q02**|378|396|378|408|381|
|**Q03**|2053|2217|2097|2068|2072|
|**Q04**|1030|1418|1058|1071|1101|
|**Q05**|2494|2573|2587|2615|2706|
|**Q06**|962|822|829|966|963|
|**Q07**|2298|2383|2426|2390|2493|
|**Q08**|1563|1630|1627|1649|1693|
|**Q09**|3503|3734|3713|3612|3775|
|**Q10**|2370|2787|2583|2533|2434|
|**Q11**|468|324|336|328|323|
|**Q12**|1037|1455|1000|1137|1017|
|**Q13**|2235|2485|2309|2290|2288|
|**Q14**|946|953|1158|1000|1013|
|**Q15**|823|1168|1336|986|1245|
|**Q16**|901|900|939|954|927|
|**Q17**|853|902|881|896|906|
|**Q18**|7369|9457|7562|7761|7717|
|**Q19**|970|1021|1291|1056|1317|
|**Q20**|1201|1218|1229|2011|1204|
|**Q21**|3309|3784|3394|3501|3487|
|**Q22**|440|487|451|450|460|
|**SUM**|40370|46263|42393|42904|42779

### sf1000

default with file cache

|           |             |          |          |           |           |
| --------- | ----------- | -------- | -------- | --------- | --------- |
| **Query** | **Default** | **X=10** | **X=50** | **X=100** | **X=500** |
| **Q01**   | 42902       | 44235    | 41353    | 41115     | 41215     |
| **Q02**   | 2560        | 2618     | 2855     | 2713      | 2629      |
| **Q03**   | 34505       | 39886    | 34458    | 34203     | 33721     |
| **Q04**   | 15230       | 20644    | 14710    | 14893     | 14707     |
| **Q05**   | 44124       | 46042    | 44598    | 44357     | 43497     |
| **Q06**   | 16158       | 15289    | 15463    | 15690     | 15519     |
| **Q07**   | 38524       | 37951    | 37997    | 38174     | 37767     |
| **Q08**   | 32042       | 36543    | 31965    | 32340     | 31386     |
| **Q09**   | 63988       | 73457    | 65299    | 64979     | 62692     |
| **Q10**   | 37215       | 43144    | 36606    | 36745     | 36222     |
| **Q11**   | 4488        | 5055     | 4539     | 4541      | 4482      |
| **Q12**   | 17615       | 20101    | 17354    | 17617     | 17249     |
| **Q13**   | 34092       | 33440    | 34711    | 34352     | 33010     |
| **Q14**   | 19409       | 20599    | 18753    | 18779     | 18579     |
| **Q15**   | 17575       | 17610    | 17121    | 17142     | 17232     |
| **Q16**   | 7379        | 8015     | 7543     | 7560      | 7335      |
| **Q17**   | 14899       | 21776    | 14447    | 14715     | 14386     |
| **Q18**   | 0           | 0        | 0        | 0         | 0         |
| **Q19**   | 22852       | 21751    | 22377    | 22565     | 22161     |
| **Q20**   | 15206       | 17359    | 14863    | 14986     | 14516     |
| **Q21**   | 51007       | 64570    | 50953    | 51491     | 49645     |
| **Q22**   | 5792        | 8080     | 5734     | 5930      | 5525      |
| **SUM**   | 537562      | 598165   | 533699   | 534887    | 523475    |

no file cache

|           |             |          |          |           |           |
| --------- | ----------- | -------- | -------- | --------- | --------- |
| **Query** | **Default** | **X=10** | **X=50** | **X=100** | **X=500** |
| **Q01**   | 42824       | 43872    | 41407    | 41565     | 41157     |
| **Q02**   | 2511        | 2597     | 2669     | 2605      | 2549      |
| **Q03**   | 34444       | 40297    | 33837    | 34215     | 33627     |
| **Q04**   | 15217       | 20507    | 14929    | 14813     | 14666     |
| **Q05**   | 44680       | 45787    | 43781    | 44366     | 43413     |
| **Q06**   | 16201       | 15202    | 15397    | 15609     | 15556     |
| **Q07**   | 38519       | 38537    | 37722    | 38051     | 37524     |
| **Q08**   | 32101       | 36635    | 31620    | 32110     | 31194     |
| **Q09**   | 64030       | 74037    | 64050    | 63648     | 62721     |
| **Q10**   | 36974       | 41708    | 36374    | 36633     | 36641     |
| **Q11**   | 4554        | 4485     | 4525     | 4500      | 4546      |
| **Q12**   | 17664       | 19880    | 17299    | 17396     | 17320     |
| **Q13**   | 33198       | 32047    | 34531    | 34543     | 32828     |
| **Q14**   | 19130       | 20479    | 18532    | 18880     | 18590     |
| **Q15**   | 17597       | 17930    | 17154    | 17336     | 17038     |
| **Q16**   | 7253        | 7887     | 7564     | 7357      | 7387      |
| **Q17**   | 15106       | 20625    | 14469    | 14576     | 14463     |
| **Q18**   | 0           | 0        | 0        | 0         | 0         |
| **Q19**   | 23098       | 22080    | 22290    | 22307     | 22440     |
| **Q20**   | 14905       | 17010    | 14969    | 15094     | 14459     |
| **Q21**   | 50878       | 64490    | 50734    | 50894     | 49500     |
| **Q22**   | 5589        | 8109     | 5845     | 5991      | 5672      |
| **SUM**   | 536473      | 594201   | 529698   | 532489    | 523291    |

### conclusion

comparing to default conf, X=10 increases about 10% latency, but X=50/100/500 does not add latency at all.

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/exec/pipeline/scan/morsel.cpp [hunk 1]
```diff
@@ -35,6 +35,26 @@ namespace starrocks::pipeline {
...
+        if (scan_range.__isset.empty && scan_range.empty) {
+            if (scan_range.__isset.has_more) {
+                *has_more_morsel = scan_range.has_more;
+            }
+            continue;
+        }
+        morsels.emplace_back(std::make_unique<pipeline::ScanMorsel>(node_id, scan_range));
+    }
+
+    if (morsels.empty() && !accept_empty_scan_ranges) {
```
Comment (stephen-shelby) [inline be/src/exec/pipeline/scan/morsel.cpp:53]: is `accept_empty_scan_ranges` inversion ?

File: be/src/exec/pipeline/fragment_executor.cpp [hunk 2]
```diff
@@ -852,4 +852,34 @@ void FragmentExecutor::_fail_cleanup(bool fragment_has_registed) {
...
+    FragmentContextPtr fragment_ctx = query_ctx->fragment_mgr()->get(instance_id);
+    if (fragment_ctx == nullptr) return Status::OK();
+
+    for (const auto& [node_id, scan_ranges] : params.per_node_scan_ranges) {
+        auto iter = fragment_ctx->morsel_queue_factories().find(node_id);
+        if (iter == fragment_ctx->morsel_queue_factories().end()) {
+            continue;
+        }
+        MorselQueueFactory* morsel_queue_factory = iter->second.get();
+        if (morsel_queue_factory == nullptr) {
```
Comment (stephen-shelby) [inline be/src/exec/pipeline/fragment_executor.cpp:873]: when is morsel_queue_factory null?

File: be/src/exec/pipeline/scan/scan_operator.cpp [hunk 3]
```diff
@@ -203,7 +207,7 @@ bool ScanOperator::is_finished() const {
     }
 
     // Any io task is running or needs to run.
-    if (_num_running_io_tasks > 0 || !_morsel_queue->empty()) {
+    if (_num_running_io_tasks > 0 || _morsel_queue->has_more() || !_morsel_queue->empty()) {
```
Comment (stephen-shelby) [inline be/src/exec/pipeline/scan/scan_operator.cpp:210]: Is it enough for `has_output()` and `is_finished()` to adapt to the `has_more()` ?

File: be/src/exec/pipeline/scan/scan_operator.cpp [hunk 4]
```diff
@@ -178,6 +178,10 @@ bool ScanOperator::has_output() const {
         }
     }
 
+    if (_morsel_queue->has_more()) {
```
Comment (stephen-shelby) [inline be/src/exec/pipeline/scan/scan_operator.cpp:181]: Is there a problem with returning false directly without check has_more?

Comment (dirtysalt) [reply be/src/exec/pipeline/scan/scan_operator.cpp]: you are correct.  it's a bug

File: be/src/exec/pipeline/scan/scan_operator.cpp (hunk 3 above)
Comment (dirtysalt) [reply be/src/exec/pipeline/scan/scan_operator.cpp:206]: yes. only `is_finished()` is enough.

File: be/src/exec/pipeline/fragment_executor.cpp (hunk 2 above)
Comment (dirtysalt) [reply be/src/exec/pipeline/fragment_executor.cpp:874]: I'm not sure.  maybe query is cancelled earlier than expected.  it's kind of like defensive code.

File: be/src/exec/pipeline/scan/morsel.cpp (hunk 1 above)
Comment (dirtysalt) [reply be/src/exec/pipeline/scan/morsel.cpp:53]: no. it's expected.

**Review by stephen-shelby — DISMISSED** [review]

File: be/src/exec/pipeline/scan/morsel.cpp [hunk 5]
```diff
@@ -35,6 +35,26 @@ namespace starrocks::pipeline {
...
 
+void ScanMorsel::build_scan_morsels(int node_id, const std::vector<TScanRangeParams>& scan_ranges,
+                                    bool accept_empty_scan_ranges, pipeline::Morsels* ptr_morsels,
+                                    bool* has_more_morsel) {
+    pipeline::Morsels& morsels = *ptr_morsels;
+    *has_more_morsel = false;
+    for (const auto& scan_range : scan_ranges) {
+        if (scan_range.__isset.empty && scan_range.empty) {
+            if (scan_range.__isset.has_more) {
+                *has_more_morsel = scan_range.has_more;
```
[the remark this answers is not in the document]
Comment (dirtysalt) [reply be/src/exec/pipeline/scan/morsel.cpp:46]: yes.

File: be/src/exec/scan_node.cpp [hunk 6]
```diff
@@ -186,15 +187,10 @@ StatusOr<pipeline::MorselQueuePtr> ScanNode::convert_scan_range_to_morsel_queue(
...
-    } else {
-        for (const auto& scan_range : scan_ranges) {
-            morsels.emplace_back(std::make_unique<pipeline::ScanMorsel>(node_id, scan_range));
-        }
-    }
-
+    [[maybe_unused]] bool has_more_morsel = false;
+    pipeline::ScanMorsel::build_scan_morsels(node_id, scan_ranges, accept_empty_scan_ranges(), &morsels,
+                                             &has_more_morsel);
+    DCHECK(has_more_morsel == false);
```
[the remark this answers is not in the document]
Comment (dirtysalt) [reply be/src/exec/scan_node.cpp:193]: for current implementation, yes. 

because only connector_scan_node.cpp supports incremental scan ranges delivery.

**Review by Youngwb — APPROVED** [review]

**Review by stephen-shelby — APPROVED** [review]

**Review by wyb — APPROVED** [review]

**Review by ZiheLiu — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Review by gengjun-git — APPROVED** [review]

**Review by wangruin — APPROVED** [review]

**Pull request merged**